Source/WTF/wtf/PreciseSum.cpp

```diff
@@ -59,123 +65,358 @@
 
 namespace WTF {
 
-static constexpr double MAX_DOUBLE = 1.79769313486231570815e+308;
-static constexpr double PENULTIMATE_DOUBLE = 1.79769313486231550856e+308;
-static constexpr double MAX_ULP = MAX_DOUBLE - PENULTIMATE_DOUBLE;
-static constexpr double TWO_POW_1023 = 8.98846567431158e+307;
+namespace {
+
+constexpr int64_t XSUM_MANTISSA_BITS = 52; // Bits in fp mantissa, excludes implict 1
+constexpr int64_t XSUM_EXP_BITS = 11; // Bits in fp exponent
+constexpr int64_t XSUM_MANTISSA_MASK = (static_cast<int64_t>(1) << XSUM_MANTISSA_BITS) - 1; // Mask for mantissa bits
+constexpr int64_t XSUM_EXP_MASK = (1 << XSUM_EXP_BITS) - 1; // Mask for exponent
+constexpr int64_t XSUM_EXP_BIAS = (1 << (XSUM_EXP_BITS - 1)) - 1; // Bias added to signed exponent
+constexpr int64_t XSUM_SIGN_BIT = XSUM_MANTISSA_BITS + XSUM_EXP_BITS; // Position of sign bit
+constexpr uint64_t XSUM_SIGN_MASK = static_cast<uint64_t>(1) << XSUM_SIGN_BIT; // Mask for sign bit
+constexpr int64_t XSUM_SCHUNK_BITS = 64; // Bits in chunk of the small accumulator
+constexpr int64_t XSUM_LOW_EXP_BITS = 5; // # of low bits of exponent, in one chunk
+constexpr int64_t XSUM_LOW_EXP_MASK = (1 << XSUM_LOW_EXP_BITS) - 1; // Mask for low-order exponent bits
+constexpr int64_t XSUM_HIGH_EXP_BITS = XSUM_EXP_BITS - XSUM_LOW_EXP_BITS; // # of high exponent bits for index
+constexpr int64_t XSUM_SCHUNKS = (1 << XSUM_HIGH_EXP_BITS) + 3; // # of chunks in small accumulator
+constexpr int64_t XSUM_LOW_MANTISSA_BITS = 1 << XSUM_LOW_EXP_BITS; // Bits in low part of mantissa
+constexpr int64_t XSUM_LOW_MANTISSA_MASK = (static_cast<int64_t>(1) << XSUM_LOW_MANTISSA_BITS) - 1; // Mask for low bits
+constexpr int64_t XSUM_SMALL_CARRY_BITS = (XSUM_SCHUNK_BITS - 1) - XSUM_MANTISSA_BITS; // Bits sums can carry into
+constexpr int64_t XSUM_SMALL_CARRY_TERMS = (1 << XSUM_SMALL_CARRY_BITS) - 1; // # terms can add before need prop.
+
+} // anonymous namespace
+
+namespace Xsum {
+
+SmallAccumulator::SmallAccumulator(
+    int addsUntilPropagate,
+    int64_t inf,
+    int64_t nan
+) : chunk(XSUM_SCHUNKS, 0LL), addsUntilPropagate { addsUntilPropagate }, inf { inf }, nan { nan } { }
 
-ALWAYS_INLINE std::pair<double, double> twosum(double x, double y)
+} // namespace Xsum
+
+PreciseSum::PreciseSum()
+: m_smallAccumulator { XSUM_SMALL_CARRY_TERMS, 0, 0 }, m_sizeCount { 0 }, m_hasPosNumber { false } { }
+
+/*
+ADD A VECTOR OF FLOATING-POINT NUMBERS TO A SMALL ACCUMULATOR. Mixes
+calls of xsumCarryPropagate with calls of xsumAdd1NoCarry.
+*/
+void PreciseSum::addList(const std::span<const double> vec)
 {
-    double hi = x + y;
-    double lo = y - (hi - x);
-    return std::make_pair(hi, lo);
+    size_t offset = 01;
```

Comment:
Is this `01` right? If it is, then use `1`.

Reply:
it should be `0`. Thanks for reviewing code and finding a typo. 
So far `addList` function isn't used, but planning to use it in a different PR for speed improvement.